mapie/regression/utils/ccp_phi_function.py

```diff
@@ -0,0 +1,974 @@
+        else:
+            self.functions = cast(NDArray, [self.functions])
+
+        if (len(self.functions) == 0) and not self.marginal_guarantee:
+            raise ValueError("You need to define the `functions` argument "
+                             "with a function or a list of functions, "
+                             "or keep marginal_guarantee argument to True.")
+
+        warn_ind: Dict[str, List[int]] = {}
+        error_ind: Dict[str, List[int]] = {}
+        for i, funct in enumerate(self.functions):
+            assert callable(funct)
+            params = inspect.signature(funct).parameters
+
+            for param, arg in params.items():
+                if (
+                    param not in ["X", "y_pred", "z"]
+                    and param != "disable_marginal_guarantee"
+                ):
+                    if arg.default is inspect.Parameter.empty:
+                        if param in error_ind:
+                            error_ind[param].append(i)
+                        else:
+                            error_ind[param] = [i]
+                    elif not isinstance(self, (PolynomialPhiFunction,
+                                               GaussianPhiFunction)):
+                        if param in warn_ind:
+                            warn_ind[param].append(i)
+                        else:
+                            warn_ind[param] = [i]
+
+        if len(warn_ind) > 0:
+            warn_msg = ""
+            for param, inds in warn_ind.items():
+                warn_msg += (
+                    f"The functions at index ({', '.join(map(str, inds))}) "
+                    + "of the 'functions' argument, has an unknown optional "
+                    + f"argument '{param}'.\n"
+                )
+            warnings.warn(
+                "WARNING: Unknown optional arguments.\n"
+                + warn_msg +
+                "The only recognized arguments are : 'X', 'y_pred' and 'z'. "
+                "The other optional arguments will act as parameters, "
+                "as it is always their default value which will be used."
+            )
+        if len(error_ind) > 0:
+            error_msg = ""
+            for param, inds in error_ind.items():
+                error_msg += (
+                    f"The functions at index ({', '.join(map(str, inds))}) "
+                    + "of the 'functions' argument, has an unknown required "
+                    + f"argument '{param}'.\n"
+                )
+            raise ValueError(
+                "Forbidden required argument.\n"
+                f"{error_msg}"
+                "The only allowed required argument are : 'X', "
+                "'y_pred' and 'z'.\n"
+                "Note: You can use optional arguments if you want "
+                "to. They will act as parameters, as it is always "
+                "their default value which will be used."
+            )
+        return cast(NDArray, self.functions)
```

Comment:
Check if you can remove the cast.

Reply:
Done, using a ``functions_`` which is the fitted, not ``None`` version of ``functions``